### PR TITLE
Issue 858: New Date entry on the new Calendar page is now based off the previous one

### PR DESCRIPTION
**Pull request by ElementUser**
# Linked Issue

https://github.com/Sendouc/sendou.ink/issues/858

- Is the correction of the following PR that I accidentally botched with some incorrect branch merges: https://github.com/Sendouc/sendou.ink/pull/1062

# Description of Changes

On the new calendar page (http://localhost:5800/calendar/new), adding a new Date entry with the "Add" button will now insert a new date with an offset of +24 hours of the previous row's date value.

## Implementation details

Added a new React Hook that keeps track of the array of DateInput components
- This is done indirectly by lifting React State up. The state handler function is passed from the DatesInput function to the DateInput function

## Manual testing
- I pressed on the Add button several times & it gave the correct +24 hour offset.
- When there are 2 or more rows, I edited the last row & pressed the Add button again & the date was offset from the last row's updated Date value correctly.
- When I removed a row and then add a row, the date was offset based on the last row's Date value correctly.

## Screenshots

image

### CODE REVIEW

File: app/components/DateInput.tsx
Comment (remmycat) [inline app/components/DateInput.tsx:27]: I think the API of the DateInput gets a little clunky like this, these two props are very specific for being used in the "DatesInput" and would possibly keep us from reusing them elsewhere.
Personally I'd prefer to have an `onChange?: (newDate: Date | null) => void` prop (or similar), where the parent can then choose what to put in there and trigger from that, depending on its own needs.

File: app/components/DateInput.tsx
Comment (remmycat) [inline app/components/DateInput.tsx:56]: One issue that already existed in this component is that there is no handling for invalid dates.
I believe most browsers will trigger an onChange with `e.target.value === ""` when the user deletes the date that we prefilled.

This then leads to `new Date("")` being put into the component state, but that's an invalid date (not throwing an error though, date objects can just be invalid).

Calling `dateToYearMonthDayHourMinuteString(date)` with an invalid date object leads to `"NaN-NaN-NaNTNaN:NaN"` 😁 

Setting the input values to these "NaN" strings will trigger a console warning by react-dom, but still "works" in that it doesn't display anything weird.

I believe we need to have the code be aware of this and test for invalid dates. One possible test you could do, to check if a date is valid is `Number.isNaN(date.getTime())`.
Where exactly is up to you, the component state could also just be the string we get from the event handler for example.

File: app/routes/calendar/new.tsx
Comment (remmycat) [inline app/routes/calendar/new.tsx:299]: This sounds like a great candidate to put into the `utils/date.ts` file where we are also importing the other utils from :) Maybe the offset could be a function parameter?

File: app/routes/calendar/new.tsx
Comment (remmycat) [inline app/routes/calendar/new.tsx:362]: I think we can save us the trouble of always keeping two states in sync, by just deleting the `datesCount` state?
I don't see a reason to keep it, when we can just as well iterate over the values inside the `datesInputState` when rendering the date inputs, and will also get rid of the typescript error here, because we don't need to read it from a value 🙂 

(`datesCount === datesInputState.length`)

**Review by remmycat — COMMENTED** [review]
Looking really good already! ✨ 

I added a few suggestions on ways we could improve the code further 🙂 

I also tried this in two browsers (Safari and Firefox), and found the same issue in both:
When deleting something from the current date, all the newly added ones are not prefilled at all, (or rather, have the value set to a chain of "NaN"s, which the browser interprets as an empty value)

I think this is mostly due to the code so far not being aware of the possibility of invalid / incomplete dates, I added a more detailed comment in the DateInput for that.

In the cases where we cannot calculate the next date because the first is invalid, my expectation would be to just have the current date again as the initial value there. Or do you think something else makes sense? If we decide on an empty value (current behaviour), we should at least do that properly by specifying `value=""` in the input

#### Screenshot of me having deleted the month from the input and then adding three new dates:
Screenshot 2022-10-30 at 01 35 30

#### The HTML:
Screenshot 2022-10-30 at 01 35 45

File: app/components/DateInput.tsx
Comment (ElementUser) [reply app/components/DateInput.tsx:27]: The problem that I have at the moment is that (as far as I am aware), the app is only using React Hooks for state management (and what we have here isn't suitable for persisting in the DB even temporarily, since we are preparing the POST form for it). 

The way we're doing it now, we're making the child responsible for updating the state of the parent, rather than the parent be responsible for updating the state of its child components.

==

Personally, I would have used a state management library to handle it rather than a series of workarounds (either context, or Hooks as far as what we have now). Without state management in some kind of form, I (currently) cannot think of a way for the parent component to become aware of the changes in the child component (or vice-versa).

As far as your current suggestion goes: let's say that I do the change you suggested & make the child update its value via an onchange callback function (which it basically already does). Currently I see 1 of 2 specific obstacles that I would need to overcome:

1. How would the parent (`DatesInput`) be aware of the changes when generating a new child component (`DateInput`)? OR:
2. How would the child (`DateInput`) update the state that the parent (`DatesInput`) has access to in order to know what to base the new date value off of when it creates a new child component

Without finding a solution to either of these issues, I don't think I can make the change you're requesting at this moment

File: app/components/DateInput.tsx
Comment (ElementUser) [reply app/components/DateInput.tsx:56]: This is fair & the initial of the focus in the PR assumed that validation was handled correctly. I can look into validating the date & coming up with a fallback when I find time later =) thanks for pointing this out!

File: app/routes/calendar/new.tsx
Comment (ElementUser) [reply app/routes/calendar/new.tsx:299]: That sounds like a good idea, I'll port it over there & add your suggestion. I only set the offset as a constant here because I did not anticipate it to be reused haha, but I will do the refactor

File: app/routes/calendar/new.tsx
Comment (ElementUser) [reply app/routes/calendar/new.tsx:362]: Yes I agree & will refactor it when the time comes

File: app/components/DateInput.tsx
Comment (remmycat) [reply app/components/DateInput.tsx:27]: While I get that we could consider the usage of a state or form management library to help with these situations, I don't think it's too bad with hooks in this case.
I'm not sure if I really understand your last two issues here, maybe I could show roughly what I mean?

So let's say we took my suggestion there, the parent component then would render the input like this. It's essentially the same code, just inside the parent.

```tsx
<DateInput
  id="date"
  name="date"
  defaultValue={
    datesInputState.at(i)?.finalDateInputDate ?? undefined
  }
  min={MIN_DATE}
  max={MAX_DATE}
  data-cy="date-input"
  required
  onChange={(newDate) => {
    setDatesInputState((current) =>
        current.map((obj) => {
          if (obj.index === i) {
            return { ...obj, finalDateInputDate: updatedDate };
          }

          return obj;
        })
      );
  }}
/>
```

The child can still be the "owner" of its own state, and only notify the parent on changes.

```tsx
<input
  id={id}
  type="datetime-local"
  value={dateToYearMonthDayHourMinuteString(date)}
  min={min ? dateToYearMonthDayHourMinuteString(min) : undefined}
  max={max ? dateToYearMonthDayHourMinuteString(max) : undefined}
  onChange={(e) => {
    const updatedDate = new Date(e.target.value);
    setDate(updatedDate);
    // parent onChange
    onChange?.(updatedDate);
  }}
  data-cy={dataCy}
  required={required}
/>
```

File: app/components/DateInput.tsx
Comment (ElementUser) [reply app/components/DateInput.tsx:56]: `Number.isNaN()` doesn't actually give me the expected result when the date object is invalid. It actually has to be the native `isNaN()` function

Comment (ElementUser) [reply app/components/DateInput.tsx]: Added getValidNewDateIfInvalid() function. This retrieves a valid date. If invalid, get a new Date object.
- So now, if we intentionally/accidentally delete the data in the Calendar's DateInput element, it will be reset to the current Date/Time

File: app/components/DateInput.tsx
Comment (ElementUser) [reply app/components/DateInput.tsx:27]: Oh okay, I didn't think about moving the Hook's setState function into the child's onChange() callback as an initializer. That's nice, I'll give it a shot - thanks!

Comment (ElementUser) [reply app/components/DateInput.tsx:27]: It works! :) that is a much cleaner way of updating the state. Thanks again!

File: app/routes/calendar/new.tsx
Comment (ElementUser) [reply app/routes/calendar/new.tsx]: Done!

**Comment by ElementUser** [comment]
@remmycat I addressed all of your suggestions & the code should be good to go now :) please let me know if there is anything else

**Review by remmycat — COMMENTED** [review]
Thank you for putting so much effort into handling my comments! 🙌

I found a few more things with the new version that I think we should address before merging it, could you check them out when you have the time?

File: app/routes/calendar/new.tsx
Comment (ElementUser) [reply app/routes/calendar/new.tsx:311]: Good idea - yeah I didn't want to touch the old code if possible, but now that we've done some refactoring, it looks like a much better idea now

File: app/routes/calendar/new.tsx
Comment (ElementUser) [reply app/routes/calendar/new.tsx:293]: Mmm, ambiguous naming in the tsx page (though that is partially on default routing mechanism) - I would have never suspected that it would use the same page when editing a pre-existing Calendar due to the name of the file.

Yes, we'll definitely have to handle this use case scenario when editing a pre-existing Calendar

File: app/utils/dates.ts
Comment (ElementUser) [reply app/utils/dates.ts:56]: Yeah you're right, I think overall the validation needs to give some kind of a visual feedback to the user if a date is invalid in the form itself.

File: app/routes/calendar/new.tsx
Comment (ElementUser) [reply app/routes/calendar/new.tsx:293]: > 1. Is the `index` relevant here, or is it enough to know the position inside the array? I feel like currently this would always look like `[{index: 0, ... }, {index: 1, ... }, {index: 2, ... }]` and so on (leaving finalDateInputDate out for brevity). I think we could get rid of this, and maybe even have the state only be of type `Array<Date>` or `Array<Date | null>`? (See later comment for why null)

Initially, I only included this because we are iterating through each object in the inputState's array & thought that we couldn't keep track of the index. I now remember that you can get the index during the iteration with the 2nd arg in `Array.map()`, so I will refactor this now & get rid of the object's index attribute :)

File: app/components/DateInput.tsx
Comment (ElementUser) [reply app/components/DateInput.tsx:52]: To be quite honest, I would have never expected a user to edit the text field directly - maybe this validation is a little much in the scope of the PR (I'll still give it a go, but there are more important things to attend to like handling the inputState initialization, now that it's been refactored).

Currently, the behaviour in the primary branch if you press backspace on the date input form field (or just deleting one of the characters in any way) will completely reset the value to the default state:

image

I'll likely just leave this as is for the moment - at the very least, when you try to Submit the form with this incorrect value, you'll get a form validation error:

image

And same if you only partially fill the date input:

image

Comment (ElementUser) [reply app/components/DateInput.tsx:52]: As far as I can tell, the consequence of **not** fixing this bug is that when you click on the `Add` button, if the last DateInput's value is invalid, then the newly generated DateInput row will also have an invalid date. So I think it's acceptable to not address this specific bug in the current PR, and separating it into a separate issue will keep the scope of the PR (and the issue it's supposed to address in the actual linked issue: https://github.com/Sendouc/sendou.ink/issues/858) well-defined

File: app/utils/dates.ts
Comment (ElementUser) [reply app/utils/dates.ts:46]: Agreed, I'll refactor this date validation function

File: app/routes/calendar/new.tsx
Comment (ElementUser) [reply app/routes/calendar/new.tsx]: #2 is the most important issue to resolve here (just leaving a note for myself here)

File: app/components/DateInput.tsx
Comment (remmycat) [reply app/components/DateInput.tsx:52]: Makes sense to me, sorry for dragging it out a bit!

Comment (remmycat) [reply app/components/DateInput.tsx:52]: Also, the screenshot that the whole value gets deleted is probably expected for your browser! The browsers just behave differently with their date inputs. But generally nowadays, all major browsers set the `value` of the input either to a valid date string, or `""` - The browser will handle the transitional states.

Comment (remmycat) [reply app/components/DateInput.tsx:52]: => #1082

File: app/routes/calendar/new.tsx
Comment (ElementUser) [reply app/routes/calendar/new.tsx]: > 2\. We need to properly initialize the state here.One gotcha is, that the route `new.tsx` is not only used for creating _new_ events, but also _edit_ existing ones. That's why the `datesCount` state got initialized with `eventToEdit?.startTimes.length ?? 1`, meaning: "If we are editing an event, use that event's data as the initial state, otherwise use 1"Currently when editing an existing event, this code will just reset the date inputs, possibly leading to data loss if the user didn't notice. We need to implement the same thing for the new state, it should initialize with the existing dates on the event, if it exists.

Okay, I handled this now too =)

Can you please have another look when you have time? Thanks again for all your feedback!

File: app/utils/dates.ts
Comment (ElementUser) [reply app/utils/dates.ts]: Will handle in https://github.com/Sendouc/sendou.ink/issues/1082

Comment (ElementUser) [reply app/utils/dates.ts]: Just to be clear, I temporarily commented out all of the error throws for invalid dates, but I still kept the console.warn() call

File: app/utils/dates.ts
Comment (ElementUser) [reply app/utils/dates.ts]: It has been refactored, though I left TODO comments for https://github.com/Sendouc/sendou.ink/issues/1082

I left the console.warn() calls for invalid dates, but commented out any thrown errors because of an invalid date

**Review by remmycat — APPROVED** [review]
Very nice job, let's merge it! 🤩  📅

**Pull request merged**
